ethstats/ethstats.go

```diff
@@ -259,81 +306,104 @@ func (s *Service) loop(ctx context.Context) {
-				if err = s.report(conn, sendCh); err != nil {
-					log.Warn("Initial stats report failed", "err", err)
-					conn.Close()
-					continue
-				}
+					// Authenticate the client with the server
+					if err = s.login(conn, sendCh); err != nil {
+						log.Warn("Stats login failed", "err", err)
+						conn.Close()
+						errTimer.Reset(connectionTimeout * time.Second)
+						continue
+					}
 
-				// Keep sending status updates until the connection breaks
-				fullReport := time.NewTicker(statusUpdateInterval * time.Second)
+					// This go routine will close when the connection gets closed/lost
+					go s.readLoop(conn)
 
-				for err == nil {
-					select {
-					case <-ctxGroup.Done():
+					// Send the initial stats so our node looks decent from the get go
+					if err = s.report(conn, sendCh); err != nil {
+						log.Warn("Initial stats report failed", "err", err)
 						conn.Close()
-						return ctxGroup.Err()
+						continue
+					}
 
-					case <-fullReport.C:
-						if err = s.report(conn, sendCh); err != nil {
-							log.Warn("Full stats report failed", "err", err)
-						}
-					case list := <-s.histCh:
-						if err = s.reportHistory(conn, list); err != nil {
-							log.Warn("Requested history report failed", "err", err)
-						}
-					case head := <-headCh:
-						if err = s.reportBlock(conn, head); err != nil {
-							log.Warn("Block stats report failed", "err", err)
-						}
-					case <-txCh:
-						if err = s.reportPending(conn); err != nil {
-							log.Warn("Transaction stats report failed", "err", err)
-						}
-					case signedMessage := <-sendCh:
-						if err = s.handleDelegateSend(conn, signedMessage); err != nil {
-							log.Warn("Delegate send failed", "err", err)
+					// Keep sending status updates until the connection breaks
+					fullReport := time.NewTicker(statusUpdateInterval * time.Second)
+
+					for err == nil {
+						select {
+						case <-ctxGroup.Done():
+							fullReport.Stop()
+							conn.Close()
+							return ctxGroup.Err()
+
+						case <-fullReport.C:
+							if err = s.report(conn, sendCh); err != nil {
+								log.Warn("Full stats report failed", "err", err)
+							}
+						case list := <-s.histCh:
+							if err = s.reportHistory(conn, list); err != nil {
+								log.Warn("Requested history report failed", "err", err)
+							}
+						case head := <-headCh:
+							if err = s.reportBlock(conn, head); err != nil {
+								log.Warn("Block stats report failed", "err", err)
+							}
+							if err = s.reportPending(conn); err != nil {
+								log.Warn("Post-block transaction stats report failed", "err", err)
+							}
+						case <-txCh:
+							if err = s.reportPending(conn); err != nil {
+								log.Warn("Transaction stats report failed", "err", err)
+							}
+						case signedMessage := <-sendCh:
+							// if it is a ping or hello message, it shouldn't be handled here
+							if signedMessage.Action != actionNodePing && signedMessage.Action != actionHello {
```

Comment:
Would it make sense to split this out into several channels (1 for each message type)? It seems like we're selecting from it in multiple places, but are only looking for a specific message.

I think with the retry logic we'll be fine, but I think it may be possible to rely on it less (but you'll probably still need it b/c of the network).

Reply:
This is actually a check that shouldn't happened because those are messages that should be consumed before, or fail due to a timeout, and that timeout should trigger a disconnect -> reconnect. But I wanted to differentiate that case for future logs if for some reason there was something under my radar.

Regardless of handling in different channels, it was something I thought about, and it's probably a thing that we could add when we change the way we send stats (to be defined in these days), because there won't be necessary for the proxy to know which message it is (if we start using it as a transport layer, instead of generator+transport as it is today)